// @n8n/config/dist/configs/expression-engine.config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpressionEngineConfig = void 0;
const zod_1 = __importDefault(require("zod"));
const decorators_1 = require("../decorators");
const expressionEngineSchema = zod_1.default.enum(['legacy', 'vm']);
let ExpressionEngineConfig = class ExpressionEngineConfig {
    constructor() {
        this.engine = 'legacy';
        this.poolSize = 1;
        this.maxCodeCacheSize = 1024;
        this.bridgeTimeout = 5000;
        this.bridgeMemoryLimit = 128;
        this.observabilityEnabled = true;
        this.tracesEnabled = true;
        this.slowEvaluationThresholdMs = 50;
        this.tracesSampleRate = 0.0;
    }
};
exports.ExpressionEngineConfig = ExpressionEngineConfig;
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE', expressionEngineSchema),
    __metadata("design:type", String)
], ExpressionEngineConfig.prototype, "engine", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_POOL_SIZE'),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "poolSize", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_MAX_CODE_CACHE_SIZE'),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "maxCodeCacheSize", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_TIMEOUT'),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "bridgeTimeout", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_MEMORY_LIMIT'),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "bridgeMemoryLimit", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_OBSERVABILITY_ENABLED'),
    __metadata("design:type", Boolean)
], ExpressionEngineConfig.prototype, "observabilityEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_TRACES_ENABLED'),
    __metadata("design:type", Boolean)
], ExpressionEngineConfig.prototype, "tracesEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_SLOW_EVAL_THRESHOLD_MS', zod_1.default.number({ coerce: true }).positive()),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "slowEvaluationThresholdMs", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_TRACES_SAMPLE_RATE', zod_1.default.number({ coerce: true }).min(0).max(1)),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "tracesSampleRate", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_EXPRESSION_ENGINE_IDLE_TIMEOUT'),
    __metadata("design:type", Number)
], ExpressionEngineConfig.prototype, "idleTimeout", void 0);
exports.ExpressionEngineConfig = ExpressionEngineConfig = __decorate([
    decorators_1.Config
], ExpressionEngineConfig);
