// @n8n/config/dist/configs/instance-settings-loader.config.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstanceSettingsLoaderConfig = void 0;
const decorators_1 = require("../decorators");
let InstanceSettingsLoaderConfig = class InstanceSettingsLoaderConfig {
    constructor() {
        this.ownerManagedByEnv = false;
        this.ownerEmail = '';
        this.ownerFirstName = 'Instance';
        this.ownerLastName = 'Owner';
        this.ownerPasswordHash = '';
        this.ssoManagedByEnv = false;
        this.ssoUserRoleProvisioning = 'disabled';
        this.oidcClientId = '';
        this.oidcClientSecret = '';
        this.oidcDiscoveryEndpoint = '';
        this.oidcLoginEnabled = false;
        this.oidcPrompt = 'select_account';
        this.oidcAcrValues = '';
        this.securityPolicyManagedByEnv = false;
        this.mfaEnforcedEnabled = false;
        this.personalSpacePublishingEnabled = true;
        this.personalSpaceSharingEnabled = true;
        this.samlMetadata = '';
        this.samlMetadataUrl = '';
        this.samlLoginEnabled = false;
        this.logStreamingManagedByEnv = false;
        this.logStreamingDestinations = '';
    }
};
exports.InstanceSettingsLoaderConfig = InstanceSettingsLoaderConfig;
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_OWNER_MANAGED_BY_ENV'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "ownerManagedByEnv", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_OWNER_EMAIL'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "ownerEmail", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_OWNER_FIRST_NAME'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "ownerFirstName", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_OWNER_LAST_NAME'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "ownerLastName", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_INSTANCE_OWNER_PASSWORD_HASH'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "ownerPasswordHash", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_MANAGED_BY_ENV'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "ssoManagedByEnv", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_USER_ROLE_PROVISIONING'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "ssoUserRoleProvisioning", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_OIDC_CLIENT_ID'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "oidcClientId", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_OIDC_CLIENT_SECRET'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "oidcClientSecret", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_OIDC_DISCOVERY_ENDPOINT'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "oidcDiscoveryEndpoint", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_OIDC_LOGIN_ENABLED'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "oidcLoginEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_OIDC_PROMPT'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "oidcPrompt", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_OIDC_ACR_VALUES'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "oidcAcrValues", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SECURITY_POLICY_MANAGED_BY_ENV'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "securityPolicyManagedByEnv", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_MFA_ENFORCED_ENABLED'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "mfaEnforcedEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_PERSONAL_SPACE_PUBLISHING_ENABLED'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "personalSpacePublishingEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_PERSONAL_SPACE_SHARING_ENABLED'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "personalSpaceSharingEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_SAML_METADATA'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "samlMetadata", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_SAML_METADATA_URL'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "samlMetadataUrl", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_SSO_SAML_LOGIN_ENABLED'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "samlLoginEnabled", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_LOG_STREAMING_MANAGED_BY_ENV'),
    __metadata("design:type", Boolean)
], InstanceSettingsLoaderConfig.prototype, "logStreamingManagedByEnv", void 0);
__decorate([
    (0, decorators_1.Env)('N8N_LOG_STREAMING_DESTINATIONS'),
    __metadata("design:type", String)
], InstanceSettingsLoaderConfig.prototype, "logStreamingDestinations", void 0);
exports.InstanceSettingsLoaderConfig = InstanceSettingsLoaderConfig = __decorate([
    decorators_1.Config
], InstanceSettingsLoaderConfig);
